infra/gravity/cluster_status.go

```diff
@@ -2,45 +2,108 @@ package gravity
+type statusValidator func(s GravityStatus) error
 
-	retry := wait.Retryer{
-		Attempts: 100,
-		Delay:    time.Second * 20,
+// a negation / reimplementation of https://github.com/gravitational/gravity/blob/7.0.0/lib/status/status.go#L180-L185
+// I'd prefer isDegraded and negation later, but I didn't find an elegant way to
+// have consumer friendly error messages when the information channel is squashed to != nil
+// -- 2020-05 walt
+func isNotDegraded(s GravityStatus) error {
+	if s.Cluster.State == constants.SiteStateDegraded {
+		return trace.Errorf("cluster state %q", s.Cluster.State)
+	}
+	if s.Cluster.SystemStatus != constants.SystemStatus_Running {
+		return trace.Errorf("expected system_status %v, found %v", constants.SystemStatus_Running, s.Cluster.SystemStatus)
 	}
+	return nil
 
-	err := retry.Do(ctx, func() error {
-		errs := make(chan error, len(nodes))
+}
 
-		for _, node := range nodes {
-			go func(n Gravity) {
-				_, err := n.Status(ctx)
-				errs <- err
-			}(node)
-		}
+// not degraded and state == active
+func isActive(s GravityStatus) error {
+	if err := isNotDegraded(s); err != nil {
+		return trace.Wrap(err)
+	}
+	if s.Cluster.State != constants.SiteStateActive {
+		return trace.Errorf("expected state %q, found %q", constants.SiteStateActive, s.Cluster.State)
+	}
+	return nil
+}
+
+func (c *TestContext) WaitForActiveStatus(nodes []Gravity) error {
+	c.Logger().WithField("nodes", Nodes(nodes)).Info("Waiting for active status.")
+	return c.WaitForStatus(nodes, isActive)
+}
+
+// Waits for all nodes to be in the same status, returns an error if nodes do not converge
+func (c *TestContext) WaitForStatus(nodes []Gravity, expected statusValidator) error {
+	b := backoff.NewExponentialBackOff()
+	b.MaxElapsedTime = c.timeouts.ClusterStatus
 
-		err := utils.CollectErrors(ctx, errs)
-		if err == nil {
-			return nil
+	expectStatus := func() (err error) {
+		statuses, err := c.Status(nodes)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		for _, sp := range statuses {
+			if sp == nil { // should never trigger, as the above err != nil check catches this path on well shaped data
+				c.Logger().WithField("statuses", statuses).Warn("Nil status pointer.")
+				return trace.Errorf("Nil status pointer.")
+			}
+			status := *sp
+			err = expected(status)
+			if err != nil {
+				c.Logger().WithError(err).WithField("status", *sp).Warn("Unexpected Status.")
+				return trace.Wrap(err)
+			}
 		}
-		c.Logger().Warnf("Status not available on some nodes, will retry: %v.", err)
-		return wait.Continue("status not ready on some nodes")
-	})
+		return nil
+	}
+
+	err := wait.RetryWithInterval(c.ctx, b, expectStatus, c.Logger())
 
 	return trace.Wrap(err)
+
+}
+
+// Status queries the status on all nodes.
+// Returns an partial status info and an error if any individual node status queries fail.
+func (c *TestContext) Status(nodes []Gravity) ([]*GravityStatus, error) {
```

Comment:
Do you need `GravityStatus` as pointers? It's more hassle if they can never be nil.